cmd/nerdctl/container/container_remove_linux_test.go

```diff
@@ -0,0 +1,125 @@
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package container
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/containerd/nerdctl/mod/tigron/expect"
+	"github.com/containerd/nerdctl/mod/tigron/require"
+	"github.com/containerd/nerdctl/mod/tigron/test"
+
+	"github.com/containerd/nerdctl/v2/pkg/rootlessutil"
+	"github.com/containerd/nerdctl/v2/pkg/testutil"
+	"github.com/containerd/nerdctl/v2/pkg/testutil/nerdtest"
+)
+
+// testContainerRmIptablesExecutor is a common executor function for testing iptables rules cleanup
+func testContainerRmIptablesExecutor(data test.Data, helpers test.Helpers) test.TestableCommand {
+	t := helpers.T()
+
+	// Get the actual container ID from the container name and store it in a label
+	containerID := data.Labels().Get("containerID")
+	if containerID == "" {
+		container := nerdtest.InspectContainer(helpers, data.Identifier())
+		containerID = container.ID
+		data.Labels().Set("containerID", containerID)
+	}
+
+	// Remove the container
+	helpers.Ensure("rm", "-f", containerID)
+
+	time.Sleep(1 * time.Second)
+
+	// Create a TestableCommand using helpers.Custom
+	if rootlessutil.IsRootless() {
+		// In rootless mode, we need to enter the rootlesskit network namespace
+		stateDir, err := rootlessutil.RootlessKitStateDir()
+		if err != nil {
+			t.Fatalf("Failed to get rootlesskit state dir: %v", err)
+		}
+
+		childPid, err := rootlessutil.RootlessKitChildPid(stateDir)
+		if err != nil {
+			t.Fatalf("Failed to get rootlesskit child pid: %v", err)
+		}
+
+		// Construct the path to the network namespace
+		uid := os.Getuid()
+		netnsPath := fmt.Sprintf("/run/user/%d/containerd-rootless/netns", uid)
+
+		if netns, err := rootlessutil.DetachedNetNS(); err != nil {
+			t.Fatalf("Failed to get detached network namespace: %v", err)
+		} else {
+			if netns != "" {
+				// First enter the user and mount namespace, then the network namespace
+				return helpers.Custom("nsenter", "-t", strconv.Itoa(childPid), "-U", "-m", "--preserve-credentials",
+					"--", "sh", "-c", fmt.Sprintf("nsenter --net=%s sh -c 'iptables -t nat -S && iptables -t filter -S && iptables -t mangle -S'", netnsPath))
+			}
+			// Enter the user and mount namespace
+			return helpers.Custom("nsenter", "-t", strconv.Itoa(childPid), "-U", "-m", "-n", "--preserve-credentials",
+				"--", "sh", "-c", "iptables -t nat -S && iptables -t filter -S && iptables -t mangle -S")
+		}
+	}
+
+	// In non-rootless mode, check iptables rules directly on the host
+	return helpers.Custom("sh", "-c", "iptables -t nat -S && iptables -t filter -S && iptables -t mangle -S")
```

Comment:
The shell script should be defined as a constant to avoid the code clone across rootless and rootful